---
title: "MACA Notebook"
output: html_notebook
---

Enter the directory of the maca folder on your drive and the name of the tissue you want to analyze.

```{r}
rootdir = "~/src/maca/"
tissue_of_interest = "Pancreas"
```

Load the requisite packages and some additional helper functions.

```{r}
library(Seurat)
library(dplyr)
library(Matrix)
```

Load the plate metadata.

```{r}
plate_metadata <- read.csv(file = paste0(rootdir, "metadata/MACA_Metadata.csv"), sep=",", header = TRUE)
colnames(plate_metadata)[1] <- "plate.barcode"
head(plate_metadata)
```

Check for available files.
```{r}
 count_files = list.files(paste0(rootdir, "data/"), pattern = "*counts.tab",recursive = FALSE)

 downloaded_plates = lapply(count_files, function(x) strsplit(x,"[./]")[[1]][1])

 plate_metadata = plate_metadata[plate_metadata$plate.barcode %in% downloaded_plates,]
 head(plate_metadata)
```

Load the plate data.

```{r}
tissue_plates = filter(plate_metadata, tissue == tissue_of_interest & mouse.age == 3)[,c('plate.barcode','tissue','subtissue','mouse.sex')]

#Load the gene names and set the metadata columns by opening the first file

raw.data = read.csv(file = paste0(rootdir,"data/", tissue_plates$plate.barcode[1], '.counts.tab') ,sep='\t')
raw.data = data.frame(row.names = rownames(raw.data))
meta.data = plate_metadata[1,][-1,]

for(i in 1:nrow(tissue_plates)){
  new.data = read.csv(file = paste0(rootdir,"data/", tissue_plates$plate.barcode[i], '.counts.tab') ,sep='\t')
  new.metadata = data.frame(row.names = colnames(new.data))
  new.metadata['plate.barcode'] = tissue_plates$plate.barcode[i]
  
  raw.data = cbind(raw.data, new.data)
  meta.data = rbind(meta.data, new.metadata)
}
rnames = row.names(meta.data)
meta.data <- merge(meta.data, plate_metadata)
row.names(meta.data) <- rnames
```


```{r}

erccs <- grep(pattern = "^ERCC-", x = rownames(x = raw.data), value = TRUE)
percent.ercc <- Matrix::colSums(raw.data[erccs, ])/Matrix::colSums(raw.data)
ercc.index <- grep(pattern = "^ERCC-", x = rownames(x = raw.data), value = FALSE)
raw.data <- raw.data[-ercc.index,]

tiss <- CreateSeuratObject(raw.data = raw.data, project = tissue_of_interest, 
                    min.cells = 5, min.genes = 5)

tiss <- AddMetaData(object = tiss, meta.data)
tiss <- AddMetaData(object = tiss, percent.ercc, col.name = "percent.ercc")
# Change default name for sums of counts from nUMI to nReads
colnames(tiss@meta.data)[colnames(tiss@meta.data) == 'nUMI'] <- 'nReads'
```


Calculate percent ribosomal genes.

```{r}
ribo.genes <- grep(pattern = "^Rp[sl][[:digit:]]", x = rownames(x = tiss@data), value = TRUE)

percent.ribo <- Matrix::colSums(tiss@raw.data[ribo.genes, ])/Matrix::colSums(tiss@raw.data)

tiss <- AddMetaData(object = tiss, metadata = percent.ribo, col.name = "percent.ribo")
```

A sanity check: reads vs genes.

```{r}
GenePlot(object = tiss, gene1 = "nReads", gene2 = "nGene")
```

Filter out cells with few reads and few genes.

```{r}
tiss <- FilterCells(object = tiss, subset.names = c("nGene", "nReads"), 
    low.thresholds = c(500, 50000), high.thresholds = c(25000, 2000000))
```


Normalize the data, then regress out correlation with total reads
```{r}
tiss <- NormalizeData(object = tiss)
tiss <- ScaleData(object = tiss, vars.to.regress = c("nReads", "percent.ribo","Rn45s"))
tiss <- FindVariableGenes(object = tiss, do.plot = TRUE, x.high.cutoff = Inf, y.cutoff = 0.5)
```


Run Principal Component Analysis.
```{r}
tiss <- RunPCA(object = tiss, pc.genes = tiss@var.genes, do.print = TRUE, pcs.print = 1:5, 
    genes.print = 5)
```

```{r}
tiss <- ProjectPCA(object = tiss, do.print = FALSE)
```


```{r, echo=FALSE, fig.height=4, fig.width=8}
PCHeatmap(object = tiss, pc.use = 1:3, cells.use = 500, do.balanced = TRUE, label.columns = FALSE, num.genes = 12)
```

```{r}
PCElbowPlot(object = tiss)
```


```{r}
tiss <- FindClusters(object = tiss, reduction.type = "pca", dims.use = 1:20, 
    resolution = 1.0, print.output = 0, save.SNN = TRUE)
```


```{r}
tiss <- RunTSNE(object = tiss, dims.use = 1:20, seed.use = 10, check_duplicates = F)
```

```{r}
# note that you can set do.label=T to help label individual clusters
TSNEPlot(object = tiss, do.label = T)
```

Check expression of genes of interset.
```{r}
genes_to_check = c('Ins1', 'Gcg', 'Ppy', 'Sst')

FeaturePlot(tiss, genes_to_check, pt.size = 1)
```

Color by metadata, like plate barcode, to check for batch effects.
```{r}
TSNEPlot(object = tiss, do.return = TRUE, group.by = "plate.barcode")
```

Find differentially expressed markers.
```{r}
tiss.markers <- FindAllMarkers(object = tiss, only.pos = TRUE, min.pct = 0.25, 
    thresh.use = 0.25)
```

Display top 4 markers per cluster.
```{r}
tiss.markers %>% group_by(cluster) %>% top_n(4, avg_diff)
```


```{r}
# To change the y-axis to show raw counts, add use.raw = T.
VlnPlot(tiss, genes_to_check)
```


```{r}
# Get markers for a particular cluster
cluster0_markers = filter(tiss.markers, cluster == 0)$gene

DotPlot(tiss, genes.plot = cluster0_markers[1:7], plot.legend = T)
```


```{r}
tiss = BuildClusterTree(tiss)
```

Get markers for a particular cluster
```{r}
cluster_markers = filter(tiss.markers, cluster == 4)$gene

DotPlot(tiss, genes.plot = cluster_markers[1:10])
```


## Subset and iterate

We can repeat the above analysis on a subset of genes, defined using cluster IDs or expression or some other metadata. This is a good way to drill down and find substructure.

```{r}
# Get subtissue, eg, the endocrine clusters
subtiss <- SubsetData(object = tiss, ident.use = c(0, 1, 12),
  do.center = F, do.scale = F)

subtiss <- NormalizeData(object = subtiss)
subtiss <- ScaleData(object = subtiss, vars.to.regress = c("nReads", "percent.ribo","Rn45s"))
```



```{r}
subtiss <- FindVariableGenes(object = subtiss, do.plot = TRUE, x.high.cutoff = Inf, y.cutoff = 0.5)
subtiss <- RunPCA(object = subtiss, pcs.compute = 20)
subtiss <- ProjectPCA(object = subtiss, do.print = FALSE)
```


```{r}
PCElbowPlot(object = subtiss)
```

```{r}
# If this fails for your subset, it may be that cells.use is more cells than you have left! Try reducing it.
PCHeatmap(object = subtiss, pc.use = 1:6, cells.use = 250, do.balanced = TRUE, label.columns = FALSE, num.genes = 12)
```

You may want to play with different resolutions. If the subset is small and you are still getting big clusters, try decreasing k.param to 10 or so.
```{r}
# You may want to play with different resolutions 
subtiss <- FindClusters(object = subtiss, reduction.type = "pca", dims.use = 1:20, 
    resolution = 1.0, print.output = 0, save.SNN = TRUE, force.recalc = T, k.param = 20)
```


```{r}
subtiss <- RunTSNE(object = subtiss, dims.use = 1:20, seed.use = 10, check_duplicates = F)
```

```{r}
# note that you can set do.label=T to help label individual clusters
TSNEPlot(object = subtiss, do.label = T)
```

Check expression of genes of interset.
```{r}
genes_to_check = c('Ins1', 'Gcg', 'Ppy', 'Sst')

FeaturePlot(subtiss, genes_to_check, pt.size = 1)
```


```{r}
GenePlot(subtiss, 'Ins1', 'Gcg', use.raw = T)
```
