---
title: "Seurat Internals"
output: html_notebook
---

This notebook is for recording details of how Seurat runs internally, so we can verify and understand normalized, scaled, regressed data.

The NormalizeData function used is:

```{r}
lognorm = function(x, s) log1p(x/sum(x) *s)
```

Verification:

```{r}
all.equal(lognorm(tiss@raw.data[,1], 1e4), tiss@data[,1])
```

By scaling before adding 1, we set a natural resolution. When x is larger than 1 part in 10,000, then the +1 doesn't matter much, and doubling the proportion of x will add a constant factor to the log.

When x is less than 1 part in 10,000, then we lose that loglinear relatinoship This is not obviously desirable to me. Or if it is, it makes much more sense for experiments where the value 10^4 is within an order of magnitude of the actual typical number of reads/UMI.

"By default, we employ a global-scaling normalization method “LogNormalize” that normalizes the gene expression measurements for each cell by the total expression, multiplies this by a scale factor (10,000 by default), and log-transforms the result."

```{r}
x = 2 ^ (1:20)
plot(lognorm(x))
```

To fix this, we can adjust:

```{r}
adjust_lognorm = function(x, rescaling) log1p(rescaling*(exp(x)-1))
```

```{r}
all.equal(adjust_lognorm(lognorm(tiss@raw.data[,1], 1e4), 1e2),lognorm(tiss@raw.data[,1], 1e6))
```



If you use "genesCLR", the function is:

```{r}
custom_function = function(x) log1p((x)/(exp(sum(log1p((x)[x > 0]), na.rm=TRUE) / length(x+1))))
```

```{r}
colnames(tiss@data)[1] == colnames(tiss@raw.data)[1]
```


```{r}
unnormed = tiss@raw.data[,1]
normed = tiss@data[,1]
lognormed = log1p(unnormed/sum(unnormed))
logged = log1p(unnormed)
logged_scaled = log1p(unnormed/sum(unnormed)*10000)
```

```{r}
plot(unnorm, normed, log = 'x')
plot(lognormed, normed)
plot(logged, normed)
plot(logged_scaled, normed)
```

```{r}
custom_function(tiss@data[,1])
```

```{r}
custom_function(tiss@raw.data[,1])
```


where `log1p` computes _log(1+x)_ accurately. 


This is an [R Markdown](http://rmarkdown.rstudio.com) Notebook. When you execute code within the notebook, the results appear beneath the code. 

Try executing this chunk by clicking the *Run* button within the chunk or by placing your cursor inside it and pressing *Cmd+Shift+Enter*. 

```{r}
plot(cars)
```

Add a new chunk by clicking the *Insert Chunk* button on the toolbar or by pressing *Cmd+Option+I*.

When you save the notebook, an HTML file containing the code and output will be saved alongside it (click the *Preview* button or press *Cmd+Shift+K* to preview the HTML file). 

The preview shows you a rendered HTML copy of the contents of the editor. Consequently, unlike *Knit*, *Preview* does not run any R code chunks. Instead, the output of the chunk when it was last run in the editor is displayed.

